keps/sig-network/4631-loadbalancerstatus/README.md

```diff
@@ -0,0 +1,1210 @@
+
+- Requiring all cloud providers to implement all existing
+  load-balancer-related features.
+
+[cloud-provider-gcp]: https://github.com/kubernetes/cloud-provider-gcp/
+[cloud-provider-kind]: https://github.com/kubernetes-sigs/cloud-provider-kind/
+
+## Proposal
+
+It seems uncontroversial that the cloud provider should provide
+feedback to the user when it is in the process of provisioning a load
+balancer, or when it has conclusively failed to do so.
+
+```
+<<[UNRESOLVED split the KEP? ]>>
+
+Should we split the presumably-uncontroversial
+`LoadBalancerProvisioning` and `LoadBalancerServing` Conditions out
+into a separate KEP?
+
+<<[/UNRESOLVED]>>
+```
+
+That leaves three problems to solve.
+
+### Expected Behavior When the Cloud Provider Knows It Can't Provide a Working Load Balancer At All
+
+A cloud provider might find itself unable to provide a load balancer
+for a Service *at all* because:
+
+  - The Service uses exclusively UDP (or SCTP) ports, but the
+    cloud/load balancer only supports TCP, so a load balancer would
+    be unable to route any traffic to the Service.
+
+  - The Service is single-stack IPv6 in a cloud (or cluster) that only
+    supports IPv4 load balancers, or vice versa, so a load balancer
+    would be unable to route any traffic to the Service.
+
+  - The Service has `AllocateLoadBalancerNodePorts: false`, but the
+    cloud only supports NodePort-based load balancing.
+
+  - The Service is `ExternalTrafficPolicy: Local` but the cloud cannot
+    implement client source IP preservation, and so it does not bother
+    to implement the `HealthCheckNodePort` aspect of local traffic
+    policy either, so if it created a load balancer, it would end up
+    randomly misrouting traffic to nodes that won't accept it.
+
+Currently there is no way for a cloud provider to report failure back
+to the user, and existing implementations vary between "do nothing
+(except log an error)" and "provide a load balancer that won't
+actually work".
+
+We should allow cloud providers to indicate to the user that they
+can't (or won't) provide a load balancer for a given service, and
+require load balancers to do that in cases where they know they can't
+provide at least a minimally-functional load balancer.
+
+```
+<<[UNRESOLVED clouds with multiple load balancer types ]>>
+
+Some cloud providers have multiple types of load balancers, with
+different types having different features. Does it make sense to
+suggest that cloud providers should automatically switch LB types if
+needed to implement the requested functionality? (For example,
+`cloud-provider-aws` currently defaults to "ELB" load balancers, which
+use a `Proxy` style and can't implement source IP preservation. Would
+it make sense for them to automatically switch to "NLB" load balancers
+if the Service is `ExternalTrafficPolicy: Local`?
+
+If so, should we add a field to `LoadBalancerStatus` (or
+`LoadBalancerIngress`) to indicate the LB type that was chosen?
```

Comment:
unfortunately the current state of Types is based on annotations, so this will not be feasible, as users create the service with annotations foo and bar, so it explicitly indicates the intention of using the loadbalancer of that kind ... if the controller goes its way and change these annotations it can also create a hotloop situation